src/TritonCTS/src/SinkAgglDriver.h

```diff
@@ -0,0 +1,253 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+namespace cts {
+
+struct node {
+  int node1, node2;
+  double dist;
+};
+
+inline bool operator< (const node a, const node b) {
+  return (a.dist < b.dist);
+}
+
+class clusterResult
+{
+private:
+  std::vector<node> Z;
+  int pos;
+
+public:
+  clusterResult(const int size)
+  {
+    Z.resize(size);
+    pos = 0;
+  }
+
+  void append(const int node1, const int node2, const double dist) {
+    Z[pos].node1 = node1;
+    Z[pos].node2 = node2;
+    Z[pos].dist  = dist;
+    ++pos;
+  }
+
+  node * operator[] (const int idx) { return &Z[idx]; }
+
+};
+
+class doublyLinkedList {
+  /*
+    Class for a doubly linked list. Initially, the list is the integer range
+    [0, size]. We provide a forward iterator and a method to delete an index
+    from the list.
+
+    Typical use: for (i=L.start; L<size; i=L.succ[I])
+    or
+    for (i=somevalue; L<size; i=L.succ[I])
+  */
+public:
+  int start;
+  std::vector<int> succ;
+
+private:
+  std::vector<int> pred;
+  // Not necessarily private, we just do not need it in this instance.
+
+public:
+  doublyLinkedList(const int size)
+    // Initialize to the given size.
+  {
+    start = 0;
+    succ.resize(size+1);
+    pred.resize(size+1);
+
+    for (int i=0; i<size; ++i) {
+      pred[i+1] = i;
+      succ[i] = i+1;
+    }
+    // pred[0] is never accessed!
+    //succ[size] is never accessed!
+  }
+
+  ~doublyLinkedList() {}
+
+  void remove(const int idx) {
+    // Remove an index from the list.
+    if (idx==start) {
+      start = succ[idx];
+    }
+    else {
+      succ[pred[idx]] = succ[idx];
+      pred[succ[idx]] = pred[idx];
+    }
+    succ[idx] = 0; // Mark as inactive
+  }
+
+  bool isInactive(int idx) const {
+    return (succ[idx]==0);
+  }
+};
+
+/*
+  Lookup function for a union-find data structure.
+
+  The function finds the root of idx by going iteratively through all
+  parent elements until a root is found. An element i is a root if
+  nodes[i] is zero. To make subsequent searches faster, the entry for
+  idx and all its parents is updated with the root element.
+ */
+class unionFind {
```

Comment:
Boost appears to support a disjoint set data structure would it work for this use case?